### PR TITLE
CUC Time parsing and integration

### PR DESCRIPTION
**Pull request by local-minimum**
* Interprets CUC time with TAI as epoch (where applies), this is a partial implementation since the general CUC format neither strictly assumes seconds as unit, nor is the fine time necessarily 2 octets but can be 0-4 octets, we require 2 as an uint16 since that's what innosat uses.

* Integrates with Telemetry header struct

**NOTE:** The specification for the coarse time says " Each octet within the T-field represents the state of 8 consecutive bits of a binary counter, cascaded with the adjacent counters, which rolls over at a modulo of 256. The time code represented by the T-field shall increase monotonically without reversion." which is quite obscure. Are we 100% sure that this is the same as casting those bytes to an uint16 as is done now?

# Tests

* Unit testing of all functions
* For the CUC fraction using example from the CUC spec to validate interpretation is correct.

### CODE REVIEW

**Review by joakimmoller — CHANGES_REQUESTED** [review]
Something is strange, could be the datasets, implementation, or specs, here are some examples:

The fields I've printed are ramses package time, sequence, payload time (CUC)
```
.\bin\cli.exe .\MyArchive2_20190301-150759_20190301-150844.rac
2019-03-01 15:07:59.7 +0000 UTC  8473 1958-01-01 02:44:36.442169189 +0000 UTC
2019-03-01 15:08:00.7 +0000 UTC  8474 1958-01-01 02:44:38.692169189 +0000 UTC
2019-03-01 15:08:01.7 +0000 UTC  8475 1958-01-01 02:44:39.305450439 +0000 UTC
2019-03-01 15:08:02.7 +0000 UTC  8476 1958-01-01 02:44:39.442169189 +0000 UTC
2019-03-01 15:08:03.7 +0000 UTC  8477 1958-01-01 02:44:40.305450439 +0000 UTC

.\bin\cli.exe .\Payload_RT_20190629-140042_20190629-142214.rac
2019-06-29 14:01:02.727 +0000 UTC 10939 1997-06-24 14:01:19.963195801 +0000 UTC
2019-06-29 14:01:02.727 +0000 UTC 10940 1997-06-24 14:01:19.431945801 +0000 UTC
2019-06-29 14:01:03.727 +0000 UTC 10941 1997-06-24 14:01:20.400695801 +0000 UTC
2019-06-29 14:01:04.727 +0000 UTC 10942 1997-06-24 14:01:21.353820801 +0000 UTC
2019-06-29 14:01:05.727 +0000 UTC 10943 1997-06-24 14:01:23.900695801 +0000 UTC
2019-06-29 14:01:06.727 +0000 UTC 10944 1997-06-24 14:01:23.275695801 +0000 UTC
2019-06-29 14:01:07.711 +0000 UTC 10945 1997-06-24 14:01:24.697570801 +0000 UTC

.\bin\cli.exe .\MyArchive1_20200323-013938_20200323-023005.rac
2020-03-23 01:39:38.458 +0000 UTC   635 1958-01-01 04:39:11.859954834 +0000 UTC
2020-03-23 01:39:38.458 +0000 UTC   636 1958-01-01 04:39:11.602142334 +0000 UTC
2020-03-23 01:39:38.458 +0000 UTC   637 1958-01-01 04:39:11.004486084 +0000 UTC
2020-03-23 01:39:38.458 +0000 UTC   638 1958-01-01 04:39:12.973236084 +0000 UTC
2020-03-23 01:39:38.458 +0000 UTC   639 1958-01-01 04:39:11.738861084 +0000 UTC
2020-03-23 01:39:38.458 +0000 UTC   640 1958-01-01 04:39:11.231048584 +0000 UTC
2020-03-23 01:39:38.458 +0000 UTC   641 1958-01-01 04:39:11.371673584 +0000 UTC
```
here's another print with interpreted CUC data from the STAT report at the end
```
2019-06-29 14:02:24.714 +0000 UTC 11056  1997-06-24 14:02:41.400695801 +0000 UTC 1959-01-28 03:46:09.517944336 +0000 UTC
2019-06-29 14:02:25.716 +0000 UTC 11057  1997-06-24 14:02:42.150695801 +0000 UTC 1959-01-28 03:46:10.517944336 +0000 UTC
2019-06-29 14:02:26.715 +0000 UTC 11058  1997-06-24 14:02:43.713195801 +0000 UTC 1959-01-28 03:46:11.517944336 +0000 UTC
```
So maybe the reference time i off (TAI, 1970-01-01 etc) maybe it is satelite time, but the the fractions of seconds jump up and down in the measurement sequence. I guess they should grow with the series.

File: internal/ccsds/cuc.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+func UnsegmentedTimeNanoseconds(coarseTime uint32, fineTime uint16) int64 {
+	var nanos int64 = int64(coarseTime) * int64(milliesPerSecond)
+	var fine float64 = 0
+	var flag uint16 = 1
+	for pos := 0; pos < 16; pos++ {
+		if (fineTime & flag) == flag {
+			fine += math.Pow(2, -float64(pos))
+		}
+		flag = flag << 1
+	}
```
Comment (joakimmoller) [inline internal/ccsds/cuc.go]: I think you read from the wrong end:

```suggestion
	var mask uint16 = 0x8000
	for pos := 0; pos < 16; pos++ {
		fine += float64(fineTime&(mask>>pos)) * math.Pow(2, -float64(pos+1))
	}

```

**Comment by local-minimum** [comment]
After some hacking around in `cli.go` I get the following output, which seems quite reasonable with regards to the `fineTime`:
```
1958-01-01 03:48:44.988494873 +0000 UTC
1958-01-01 03:48:44.990325928 +0000 UTC
1958-01-01 03:48:44.992156982 +0000 UTC
1958-01-01 03:48:44.994018555 +0000 UTC
1958-01-01 03:48:44.995849609 +0000 UTC
1958-01-01 03:48:44.997711182 +0000 UTC
1958-01-01 03:48:44.999572754 +0000 UTC
1958-01-01 03:48:45.00189209 +0000 UTC
1958-01-01 03:48:45.00378418 +0000 UTC
1958-01-01 03:48:45.005645752 +0000 UTC
1958-01-01 03:48:45.007476807 +0000 UTC
1958-01-01 03:48:45.009338379 +0000 UTC
1958-01-01 03:48:45.011199951 +0000 UTC
1958-01-01 03:48:45.013031006 +0000 UTC
1958-01-01 03:48:45.014892578 +0000 UTC
```

**Comment by local-minimum** [comment]
The year bit can either be that the seconds integer is not correctly parsed or that the reference time is wrong. Unfortunately there's no example in the spec about the coarse time and there's no explicit reference to which epoch innosat uses. But we also just have test-data and it could well be it starts at 0 just because it was easier that way.

File: internal/ccsds/cuc.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// in the specification.
+//
+// * coarseTime is time since epoch in seconds
+// * fineTime is subseconds is a binary division of a second
+func UnsegmentedTimeNanoseconds(coarseTime uint32, fineTime uint16) int64 {
+	var nanos int64 = int64(coarseTime) * int64(milliesPerSecond)
+	var fine = big.NewFloat(float64(fineTime))
+	fine.SetMantExp(fine, -15)
+	fineValue, _ := fine.Float64()
+	return nanos + int64(math.Round(fineValue*milliesPerSecond))
```
Comment (joakimmoller) [inline internal/ccsds/cuc.go:30]: Perhaps more correct to use https://golang.org/pkg/math/big/#Float.Int64
```suggestion
	fineValue, _ := fine.Int64()
	return nanos + int64(fineValue*milliesPerSecond)

```

**Review by joakimmoller — APPROVED** [review]
OK for now. Let's go back to the details when we know more.

File: internal/ccsds/cuc.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// "Time Code Formats" (CCSDS 301.0-B-2 - CCSDS 301.0-B-4).
+package ccsds
+
+import (
+	"math"
+	"math/big"
+	"time"
+)
+
+var milliesPerSecond float64 = math.Pow10(9)
```
Comment (joakimmoller) [inline internal/ccsds/cuc.go:11]: This should be an integer
```suggestion
var milliesPerSecond int64 = 1e9
```

File: internal/ccsds/cuc.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+// and returns the nanoseconds since epoch.
+//
+// It assumes the CUC Time follows the recommendation of having seconds as
+// units and follows the CCSDS 301.0-B-2 - CCSDS 301.0-B-4 specification
+// in the specification.
+//
+// * coarseTime is time since epoch in seconds
+// * fineTime is subseconds is a binary division of a second
+func UnsegmentedTimeNanoseconds(coarseTime uint32, fineTime uint16) int64 {
+	var nanos int64 = int64(coarseTime) * int64(milliesPerSecond)
```
Comment (joakimmoller) [inline internal/ccsds/cuc.go:26]:
```suggestion
	var nanos int64 = int64(coarseTime) * milliesPerSecond
```

**Review by joakimmoller — CHANGES_REQUESTED** [review]
Actually...

**Review by joakimmoller — APPROVED** [review]
... never mind :)

**Review by joakimmoller — APPROVED** [review]

**Pull request merged**